# Neural network using torch

## Importing required libraries

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.autograd import Variable

## Training settings

In [3]:
batch_size = 64

## Getting MNIST dataset

In [4]:
train_dataset = datasets.MNIST(root='./mnist_data/',
                               train=True,
                               transform=transforms.ToTensor(),
                               download=True)

test_dataset = datasets.MNIST(root='./mnist_data/',
                              train=False,
                              transform=transforms.ToTensor())

## Data Loader (Input Pipeline)

In [5]:
train_loader = torch.utils.data.DataLoader(dataset=train_dataset,
                                           batch_size=batch_size,
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset,
                                          batch_size=batch_size,
                                          shuffle=False)

## 1-Setting neural network model

In [6]:
class Net(nn.Module):

    def __init__(self):
        super(Net, self).__init__()
        self.l1 = nn.Linear(784, 520) # input x is 784 so we start from here
        self.l2 = nn.Linear(520, 320)
        self.l3 = nn.Linear(320, 240)
        self.l4 = nn.Linear(240, 120)
        self.l5 = nn.Linear(120, 10) # ouput is of ten categories in number. So we end it with 10

    def forward(self, x):
        x = x.view(-1, 784)  # Flatten the data (n, 1, 28, 28)-> (n, 784)
# 28*28 is 784. So, we divide equally. 1 represent color
        x = F.relu(self.l1(x))
        x = F.relu(self.l2(x))
        x = F.relu(self.l3(x))
        x = F.relu(self.l4(x))
        return self.l5(x)

In [7]:
model = Net()

## 2-Construct loss and optimizer

In [8]:
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.5)

## 3-Training Cycle
 - Forward
 - backward
 - update

In [9]:
def train(epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = Variable(data), Variable(target)
        optimizer.zero_grad()
        output = model(data)
        loss = criterion(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % 10 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))


## 4- Testing the model

In [10]:
def test():
    model.eval()
    test_loss = 0
    correct = 0
    for data, target in test_loader:
        data, target = Variable(data, volatile=True), Variable(target)
        output = model(data)
        # sum up batch loss
        test_loss += criterion(output, target).item()
        # get the index of the max
        pred = output.data.max(1, keepdim=True)[1]
        correct += pred.eq(target.data.view_as(pred)).cpu().sum()

    test_loss /= len(test_loader.dataset)
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


In [11]:
#Full range of train and test set where we can see how loss function of our prediction has changed
for epoch in range(1, 10):
    train(epoch)
    test()

Train Epoch: 1 [0/60000 (0%)]	Loss: 2.305357
Train Epoch: 1 [640/60000 (1%)]	Loss: 2.303264
Train Epoch: 1 [1280/60000 (2%)]	Loss: 2.305445
Train Epoch: 1 [1920/60000 (3%)]	Loss: 2.292897
Train Epoch: 1 [2560/60000 (4%)]	Loss: 2.290575
Train Epoch: 1 [3200/60000 (5%)]	Loss: 2.296599
Train Epoch: 1 [3840/60000 (6%)]	Loss: 2.305997
Train Epoch: 1 [4480/60000 (7%)]	Loss: 2.294274
Train Epoch: 1 [5120/60000 (9%)]	Loss: 2.298459
Train Epoch: 1 [5760/60000 (10%)]	Loss: 2.298113
Train Epoch: 1 [6400/60000 (11%)]	Loss: 2.297287
Train Epoch: 1 [7040/60000 (12%)]	Loss: 2.300120
Train Epoch: 1 [7680/60000 (13%)]	Loss: 2.305649
Train Epoch: 1 [8320/60000 (14%)]	Loss: 2.298032
Train Epoch: 1 [8960/60000 (15%)]	Loss: 2.297963
Train Epoch: 1 [9600/60000 (16%)]	Loss: 2.301248
Train Epoch: 1 [10240/60000 (17%)]	Loss: 2.309659
Train Epoch: 1 [10880/60000 (18%)]	Loss: 2.296517
Train Epoch: 1 [11520/60000 (19%)]	Loss: 2.296406
Train Epoch: 1 [12160/60000 (20%)]	Loss: 2.303596
Train Epoch: 1 [12800/60000 (

Train Epoch: 2 [44160/60000 (74%)]	Loss: 0.584893
Train Epoch: 2 [44800/60000 (75%)]	Loss: 0.393734
Train Epoch: 2 [45440/60000 (76%)]	Loss: 0.497024
Train Epoch: 2 [46080/60000 (77%)]	Loss: 0.498648
Train Epoch: 2 [46720/60000 (78%)]	Loss: 0.585934
Train Epoch: 2 [47360/60000 (79%)]	Loss: 0.469753
Train Epoch: 2 [48000/60000 (80%)]	Loss: 0.502813
Train Epoch: 2 [48640/60000 (81%)]	Loss: 0.365149
Train Epoch: 2 [49280/60000 (82%)]	Loss: 0.599904
Train Epoch: 2 [49920/60000 (83%)]	Loss: 0.433933
Train Epoch: 2 [50560/60000 (84%)]	Loss: 0.614350
Train Epoch: 2 [51200/60000 (85%)]	Loss: 0.330322
Train Epoch: 2 [51840/60000 (86%)]	Loss: 0.515252
Train Epoch: 2 [52480/60000 (87%)]	Loss: 0.593445
Train Epoch: 2 [53120/60000 (88%)]	Loss: 0.480390
Train Epoch: 2 [53760/60000 (90%)]	Loss: 0.421537
Train Epoch: 2 [54400/60000 (91%)]	Loss: 0.409210
Train Epoch: 2 [55040/60000 (92%)]	Loss: 0.399121
Train Epoch: 2 [55680/60000 (93%)]	Loss: 0.385657
Train Epoch: 2 [56320/60000 (94%)]	Loss: 0.409822


Train Epoch: 4 [27520/60000 (46%)]	Loss: 0.349989
Train Epoch: 4 [28160/60000 (47%)]	Loss: 0.243630
Train Epoch: 4 [28800/60000 (48%)]	Loss: 0.413293
Train Epoch: 4 [29440/60000 (49%)]	Loss: 0.282466
Train Epoch: 4 [30080/60000 (50%)]	Loss: 0.176570
Train Epoch: 4 [30720/60000 (51%)]	Loss: 0.353755
Train Epoch: 4 [31360/60000 (52%)]	Loss: 0.213011
Train Epoch: 4 [32000/60000 (53%)]	Loss: 0.156958
Train Epoch: 4 [32640/60000 (54%)]	Loss: 0.269601
Train Epoch: 4 [33280/60000 (55%)]	Loss: 0.597897
Train Epoch: 4 [33920/60000 (57%)]	Loss: 0.179658
Train Epoch: 4 [34560/60000 (58%)]	Loss: 0.173126
Train Epoch: 4 [35200/60000 (59%)]	Loss: 0.312791
Train Epoch: 4 [35840/60000 (60%)]	Loss: 0.278252
Train Epoch: 4 [36480/60000 (61%)]	Loss: 0.305486
Train Epoch: 4 [37120/60000 (62%)]	Loss: 0.354731
Train Epoch: 4 [37760/60000 (63%)]	Loss: 0.490716
Train Epoch: 4 [38400/60000 (64%)]	Loss: 0.194675
Train Epoch: 4 [39040/60000 (65%)]	Loss: 0.218028
Train Epoch: 4 [39680/60000 (66%)]	Loss: 0.122551


Train Epoch: 6 [10880/60000 (18%)]	Loss: 0.181825
Train Epoch: 6 [11520/60000 (19%)]	Loss: 0.234755
Train Epoch: 6 [12160/60000 (20%)]	Loss: 0.166420
Train Epoch: 6 [12800/60000 (21%)]	Loss: 0.252132
Train Epoch: 6 [13440/60000 (22%)]	Loss: 0.295558
Train Epoch: 6 [14080/60000 (23%)]	Loss: 0.064682
Train Epoch: 6 [14720/60000 (25%)]	Loss: 0.098865
Train Epoch: 6 [15360/60000 (26%)]	Loss: 0.101003
Train Epoch: 6 [16000/60000 (27%)]	Loss: 0.243807
Train Epoch: 6 [16640/60000 (28%)]	Loss: 0.096038
Train Epoch: 6 [17280/60000 (29%)]	Loss: 0.218934
Train Epoch: 6 [17920/60000 (30%)]	Loss: 0.134372
Train Epoch: 6 [18560/60000 (31%)]	Loss: 0.282329
Train Epoch: 6 [19200/60000 (32%)]	Loss: 0.232829
Train Epoch: 6 [19840/60000 (33%)]	Loss: 0.080118
Train Epoch: 6 [20480/60000 (34%)]	Loss: 0.280921
Train Epoch: 6 [21120/60000 (35%)]	Loss: 0.154970
Train Epoch: 6 [21760/60000 (36%)]	Loss: 0.095099
Train Epoch: 6 [22400/60000 (37%)]	Loss: 0.294447
Train Epoch: 6 [23040/60000 (38%)]	Loss: 0.201965


Train Epoch: 7 [55040/60000 (92%)]	Loss: 0.096327
Train Epoch: 7 [55680/60000 (93%)]	Loss: 0.170070
Train Epoch: 7 [56320/60000 (94%)]	Loss: 0.081533
Train Epoch: 7 [56960/60000 (95%)]	Loss: 0.091001
Train Epoch: 7 [57600/60000 (96%)]	Loss: 0.049777
Train Epoch: 7 [58240/60000 (97%)]	Loss: 0.047997
Train Epoch: 7 [58880/60000 (98%)]	Loss: 0.071046
Train Epoch: 7 [59520/60000 (99%)]	Loss: 0.116501

Test set: Average loss: 0.0022, Accuracy: 9571/10000 (95%)

Train Epoch: 8 [0/60000 (0%)]	Loss: 0.081855
Train Epoch: 8 [640/60000 (1%)]	Loss: 0.075526
Train Epoch: 8 [1280/60000 (2%)]	Loss: 0.080505
Train Epoch: 8 [1920/60000 (3%)]	Loss: 0.112640
Train Epoch: 8 [2560/60000 (4%)]	Loss: 0.143658
Train Epoch: 8 [3200/60000 (5%)]	Loss: 0.054374
Train Epoch: 8 [3840/60000 (6%)]	Loss: 0.068621
Train Epoch: 8 [4480/60000 (7%)]	Loss: 0.095964
Train Epoch: 8 [5120/60000 (9%)]	Loss: 0.070904
Train Epoch: 8 [5760/60000 (10%)]	Loss: 0.074624
Train Epoch: 8 [6400/60000 (11%)]	Loss: 0.056607
Train Epoch: 

Train Epoch: 9 [38400/60000 (64%)]	Loss: 0.053731
Train Epoch: 9 [39040/60000 (65%)]	Loss: 0.022243
Train Epoch: 9 [39680/60000 (66%)]	Loss: 0.085309
Train Epoch: 9 [40320/60000 (67%)]	Loss: 0.145886
Train Epoch: 9 [40960/60000 (68%)]	Loss: 0.036704
Train Epoch: 9 [41600/60000 (69%)]	Loss: 0.025163
Train Epoch: 9 [42240/60000 (70%)]	Loss: 0.043060
Train Epoch: 9 [42880/60000 (71%)]	Loss: 0.085548
Train Epoch: 9 [43520/60000 (72%)]	Loss: 0.095233
Train Epoch: 9 [44160/60000 (74%)]	Loss: 0.069109
Train Epoch: 9 [44800/60000 (75%)]	Loss: 0.068105
Train Epoch: 9 [45440/60000 (76%)]	Loss: 0.124248
Train Epoch: 9 [46080/60000 (77%)]	Loss: 0.127583
Train Epoch: 9 [46720/60000 (78%)]	Loss: 0.015866
Train Epoch: 9 [47360/60000 (79%)]	Loss: 0.070774
Train Epoch: 9 [48000/60000 (80%)]	Loss: 0.025366
Train Epoch: 9 [48640/60000 (81%)]	Loss: 0.025191
Train Epoch: 9 [49280/60000 (82%)]	Loss: 0.066547
Train Epoch: 9 [49920/60000 (83%)]	Loss: 0.112177
Train Epoch: 9 [50560/60000 (84%)]	Loss: 0.150184


**At final setp, we manage to have a 96% accuracy and 0.0017 average loss shown from our test set**